src/SymfonyCache/TaggableStore.php

```diff
@@ -0,0 +1,493 @@
+            return false;
+        }
+    }
+
+    /**
+     * @param Request $request
+     *
+     * @return string
+     */
+    public function getCacheKey(Request $request)
+    {
+        // Strip scheme to treat https and http the same
+        $uri = $request->getUri();
+        $uri = substr($uri, strlen($request->getScheme().'://'));
+
+        return 'md'.hash('sha256', $uri);
+    }
+
+    /**
+     * @param array     $vary
+     * @param HeaderBag $headerBag
+     *
+     * @return string
+     */
+    public function getVaryKey(array $vary, HeaderBag $headerBag)
+    {
+        if (0 === count($vary)) {
+            return self::NON_VARYING_KEY;
+        }
+
+        sort($vary);
+
+        $hashData = '';
+
+        foreach ($vary as $headerName) {
+            $hashData .= $headerName.':'.$headerBag->get($headerName);
+        }
+
+        return hash('sha256', $hashData);
+    }
+
+    /**
+     * @param Response $response
+     *
+     * @return string
+     */
+    public function generateContentDigest(Response $response)
+    {
+        return 'en'.hash('sha256', $response->getContent());
+    }
+
+    /**
+     * @param string $key
+     * @param string $data
+     * @param int    $expiresAfter
+     * @param array  $tags
+     *
+     * @return bool
+     */
+    private function saveDeferred($key, $data, $expiresAfter = null, $tags = [])
+    {
+        $item = $this->cache->getItem($key);
+        $item->set($data);
+        $item->expiresAfter($expiresAfter);
+
+        if (0 !== count($tags)) {
+            $item->tag($tags);
+        }
+
+        return $this->cache->saveDeferred($item);
+    }
+
+    /**
+     * Restores a Response from the cached data.
+     *
+     * @param array $cacheData An array containing the cache data
+     *
+     * @return Response|null
+     */
+    private function restoreResponse(array $cacheData)
+    {
+        $body = null;
+
+        if (isset($cacheData['headers']['x-content-digest'][0])) {
+            $item = $this->cache->getItem($cacheData['headers']['x-content-digest'][0]);
+            if ($item->isHit()) {
+                $body = $item->get();
+            }
+        }
+
+        return new Response(
+            $body,
+            $cacheData['status'],
+            $cacheData['headers']
+        );
+    }
+
+    /**
+     * @param string $cacheDir
+     *
+     * @return \Symfony\Component\Lock\StoreInterface
+     */
+    private function getBestLocalLockStore($cacheDir)
+    {
+        if (SemaphoreStore::isSupported(false)) {
+            return new SemaphoreStore();
+        } else {
+            return new FlockStore($cacheDir);
+        }
+    }
+
+    /**
+     * Increases a counter every time an item is stored to the cache and then
+     * prunes expired cache entries if a configurable threshold is reached.
+     * This only happens during write operations so cache retrieval is not
+     * slowed down.
+     */
+    private function pruneExpiredEntries()
```

Comment:
good idea! i think its ok to have this active by default. but it should be possible to disable this by setting the threshold to -1, in which case we would not even get the cache item with the counter.

we should think of some way to trigger cache pruning out of band, like in a symfony command or such. that would be the better implementation if people really care about performance.

Reply:
I chose `0` instead of `-1` but it's now possible :-)